Automatically remove stale remote jobs (7.x)

nagyo opened the pull request.

nagyo committed:
fix(cis): automatically remove successful remote jobs used for ...

... bulk ID actions

apeteri commented on cis/com.b2international.snowowl.snomed.cis.rest/src/com/b2international/snowowl/snomed/cis/rest/CisBulkSctIdService.java:
```diff
@@ -230,6 +230,7 @@
 				.setDescription(jobDescription)
 				.setRequest(request)
 				.setUser(User.SYSTEM.getUsername())
+				.setAutoClean(true)
```
Note that jobs that run into a problem will not be removed: https://github.com/b2ihealthcare/snow-owl/blob/f602f1addc56dd466941121c77c3c88768022a0b/core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/jobs/RemoteJob.java#L102-L123

On the `CisBulkJobService` CIS API level, we are offering to return jobs with a `finished` status as well, these will not be available if they are auto-cleaned. The service relies directly on the remote job storage. In case of a bulk generate job, this means the resulting SCTIDs are not available (maybe for registers this can be done because we already know what is being registered).

apeteri replied:
Although someone who monitors progress by polling the CIS bulk jobs API will not see when their registration job finishes. 😕

cmark replied:
Yeah agree, this is unfortunately against the current CIS API design, so as @nagyo already did, I suggest reverting the changes and using a better alternative like explicit clean-up tasks or GC style just in time cleanups (when a new request comes in the system will try to detect and clean any obsolete remote jobs automatically).

nagyo committed:
Revert "fix(cis): automatically remove successful remote jobs used for ..."

This reverts commit eeefa76b5692ceb773a1b196e60b36c93db54e8b.

nagyo committed:
SO-4966: automatically remove remote jobs older than the specified ...

... threshold

- remote jobs in state: FINISHED, FAILED or CANCELED will be considered
during removal

https://snowowl.atlassian.net/browse/SO-4966

nagyo committed:
SO-4966: Fix compile error, add basic unit test

https://snowowl.atlassian.net/browse/SO-4966

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into issue/SO-4966_auto_remove_bulk_id_action_jobs_7x

nagyo committed:
fix(snomed): increase test execution time to cover stale job removal

https://snowowl.atlassian.net/browse/SO-4966

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/jobs/RemoteJobTracker.java:
```diff
@@ -63,14 +70,18 @@ public void run() {
 								Expressions.builder()
-									.filter(RemoteJobEntry.Expressions.deleted(true))
-									.filter(RemoteJobEntry.Expressions.done())
+									.must(RemoteJobEntry.Expressions.done())
+									.must(Expressions.builder()
```
`filter` is more efficient than `must`; the semantics are the same, but it will not score documents based on the sub-expressions, and we don't need to keep score here.

apeteri replied:
@nagyo pointed out that `must` clauses turn into `filter`s within `EsQueryBuilder` if there are no text matching queries or script scoring (in which case the `needsScoring` variable in line 280 in the first snippet remains `false`): 

https://github.com/b2ihealthcare/snow-owl/blob/311c96ec555a64e7762c8dd34c76ce0b35e66ee3/commons/com.b2international.index/src/com/b2international/index/es/query/EsQueryBuilder.java#L232-L283

https://github.com/b2ihealthcare/snow-owl/blob/4712e259a6dd22adb06f392da6ca0549520977ae/commons/com.b2international.index/src/com/b2international/index/es/query/EsQueryBuilder.java#L174-L186

nagyo replied:
https://github.com/b2ihealthcare/snow-owl/blob/246460bb004abb6caaa6f3d474fbf46c659166ca/commons/com.b2international.index/src/com/b2international/index/es/query/EsQueryBuilder.java#L180-L185

This piece does the magic of converting musts to filters, in case no score is needed, so the end result will be a filter 🎉

apeteri reviewed (approved):
👍. I am hesitant to recommend changing `must`s to `filter`s; the functionality is unlikely to break in the future, on the other hand it might spare come cycles on the (human) reader's end if they don't know that `EsQueryBuilder` handles boolean clauses a bit differently from ES or Lucene.

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into issue/SO-4966_auto_remove_bulk_id_action_jobs_7x

nagyo committed:
feat(jobs): introduce threshold based remote job cleanup

- eliminate timer thread
- purge stale jobs

https://snowowl.atlassian.net/browse/SO-4966

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/jobs/RemoteJobTracker.java:
```diff
@@ -310,7 +344,11 @@ public void done(IJobChangeEvent event) {
+				
 				update(jobId, RemoteJobEntry.WITH_DONE, params.build());
+				
+				LOG.trace("Incrementing completed job counter to {}", completedJobCounter.incrementAndGet());
```
Please separate the counter increment to a separate line; it is easy to mistake this logging statement as a side-effect free operation:

```suggestion
				final long completedJobs = completedJobCounter.incrementAndGet();
				LOG.trace("Incrementing completed job counter to {}", completedJobs);
```

apeteri reviewed (approved):
👍 with a tiny request below.

nagyo committed:
fix(jobs): extract value of job counter to variable

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/jobs/RemoteJobTracker.java:
```diff
@@ -203,11 +168,84 @@ private void update(String jobId, String script, Map<String, Object> params) {
+			return null;
+		});
+		
+		completedJobCounter.set(getNumberOfCompletedJobs());
```
The counter here should be set to `0` and used for counting the number of completed jobs, then triggering the purge when at least X (configurable threshold) jobs have completed since the last purge.

nagyo committed:
fix(jobs): use simple job counter instead of using the number of ...

... completed jobs

- trigger remote job purge when a given job is completed instead of
scheduled
- send remote job notification for stale job removals
- update unit tests accordingly

AAAlinaaa reviewed (approved):
LGTM

cmark reviewed (approved):
LGTM!

apeteri commented:
The 8.x counterpart of this PR is https://github.com/b2ihealthcare/snow-owl/pull/942.
